File: tree/b+tree/bplus_tree.go
/*
 * B+树 - Go语言实现
 */
package main

import (
	"fmt"
)

const ORDER = 3 // B+树阶数

// BPlusNode B+树节点
type BPlusNode struct {
	keys     []int
	children []*BPlusNode
	next     *BPlusNode
	isLeaf   bool
}

// NewBPlusNode 创建新节点
func NewBPlusNode(isLeaf bool) *BPlusNode {
	return &BPlusNode{
		keys:     make([]int, 0),
		children: make([]*BPlusNode, 0),
		next:     nil,
		isLeaf:   isLeaf,
	}
}

// BPlusTree B+树结构
type BPlusTree struct {
	root *BPlusNode
}

// NewBPlusTree 创建B+树
func NewBPlusTree() *BPlusTree {
	return &BPlusTree{
		root: NewBPlusNode(true),
	}
}

// FindLeaf 查找叶节点
func (bpt *BPlusTree) FindLeaf(node *BPlusNode, key int) *BPlusNode {
	if node.isLeaf {
		return node
	}

	i := 0
	for i < len(node.keys) && key > node.keys[i] {
		i++
	}

	return bpt.FindLeaf(node.children[i], key)
}

// Search 搜索键值
func (bpt *BPlusTree) Search(key int) bool {
	leaf := bpt.FindLeaf(bpt.root, key)
	for _, k := range leaf.keys {
		if k == key {
			return true
		}
	}
	return false
}

// SplitChild 分裂子节点
func (bpt *BPlusTree) SplitChild(parent *BPlusNode, i int, child *BPlusNode) {
	newChild := NewBPlusNode(child.isLeaf)
	mid := ORDER - 1

	// 复制后半部分键
	for j := 0; j < ORDER-1; j++ {
		newChild.keys = append(newChild.keys, child.keys[j+mid+1])
	}

	// 处理子节点或链表指针
	if !child.isLeaf {
		for j := 0; j < ORDER; j++ {
			newChild.children = append(newChild.children, child.children[j+mid+1])
		}
	} else {
		newChild.next = child.next
		child.next = newChild
	}

	// 在父节点中插入中间键
	parent.keys = append(parent.keys[:i], append([]int{child.keys[mid]}, parent.keys[i:]...)...)
	parent.children = append(parent.children[:i+1], append([]*BPlusNode{newChild}, parent.children[i+1:]...)...)
}

// InsertNonFull 在非满节点中插入
func (bpt *BPlusTree) InsertNonFull(node *BPlusNode, key int) {
	if node.isLeaf {
		// 叶节点直接插入
		i := len(node.keys) - 1
		node.keys = append(node.keys, 0)
		for i >= 0 && key < node.keys[i] {
			node.keys[i+1] = node.keys[i]
			i--
		}
		node.keys[i+1] = key
	} else {
		// 内部节点
		i := len(node.keys) - 1
		for i >= 0 && key < node.keys[i] {
			i--
		}
		i++

		if len(node.children[i].keys) == 2*ORDER-1 {
			bpt.SplitChild(node, i, node.children[i])
			if key > node.keys[i] {
				i++
			}
		}
		bpt.InsertNonFull(node.children[i], key)
	}
}

// Insert 插入键值
func (bpt *BPlusTree) Insert(key int) {
	if len(bpt.root.keys) == 2*ORDER-1 {
		newRoot := NewBPlusNode(false)
		newRoot.children = append(newRoot.children, bpt.root)
		bpt.SplitChild(newRoot, 0, bpt.root)
		bpt.root = newRoot
	}
	bpt.InsertNonFull(bpt.root, key)
}

// Traverse 中序遍历
func (bpt *BPlusTree) Traverse(node *BPlusNode, result *[]int) {
	if node.isLeaf {
		*result = append(*result, node.keys...)
	} else {
		for i := 0; i < len(node.keys); i++ {
			bpt.Traverse(node.children[i], result)
			*result = append(*result, node.keys[i])
		}
		bpt.Traverse(node.children[len(node.keys)], result)
	}
}

func main() {
	fmt.Println("=" + string(make([]byte, 50)))
	fmt.Println("B+树实现 (B+ Tree)")
	fmt.Println("=" + string(make([]byte, 50)))

	bpt := NewBPlusTree()

	values := []int{10, 20, 5, 6, 12, 30, 7, 17}
	fmt.Println("\n插入数据:")
	for _, val := range values {
		fmt.Printf("%d ", val)
		bpt.Insert(val)
	}
	fmt.Println()

	result := []int{}
	bpt.Traverse(bpt.root, &result)
	fmt.Println("\n中序遍历结果:", result)

	fmt.Println("\n搜索测试:")
	testKeys := []int{6, 15, 30}
	for _, key := range testKeys {
		found := bpt.Search(key)
		if found {
			fmt.Printf("  查找 %d: 找到\n", key)
		} else {
			fmt.Printf("  查找 %d: 未找到\n", key)
		}
	}

	fmt.Println("\nB+树特点:")
	fmt.Println("  • 所有数据存储在叶节点")
	fmt.Println("  • 叶节点形成有序链表")
	fmt.Println("  • 适合磁盘存储和范围查询")
	fmt.Println("  • 查找、插入、删除: O(log n)")
	fmt.Println("  • 应用于数据库索引、文件系统")
}

File: tree/b+tree/bplus_tree.py
"""
B+树 - 多路搜索树实现
支持高效的范围查询和磁盘存储
"""


class BPlusTreeNode:
    """B+树节点"""
    def __init__(self, is_leaf=True):
        self.is_leaf = is_leaf
        self.keys = []
        self.children = []
        self.next = None  # 叶节点链表指针


class BPlusTree:
    """B+树实现 (阶数为3的简单版本)"""
    
    def __init__(self, order=3):
        self.order = order
        self.root = BPlusTreeNode()
        self.root.is_leaf = True
    
    def insert(self, key):
        """插入键值"""
        root = self.root
        
        if len(root.keys) == (2 * self.order) - 1:
            new_root = BPlusTreeNode(is_leaf=False)
            new_root.children.append(self.root)
            self._split_child(new_root, 0)
            self.root = new_root
        
        self._insert_non_full(self.root, key)
    
    def _insert_non_full(self, node, key):
        """在非满节点中插入"""
        if node.is_leaf:
            # 叶节点直接插入
            i = len(node.keys) - 1
            node.keys.append(None)
            while i >= 0 and key < node.keys[i]:
                node.keys[i + 1] = node.keys[i]
                i -= 1
            node.keys[i + 1] = key
        else:
            # 内部节点，找到合适的子节点
            i = len(node.keys) - 1
            while i >= 0 and key < node.keys[i]:
                i -= 1
            i += 1
            
            if len(node.children[i].keys) == (2 * self.order) - 1:
                self._split_child(node, i)
                if key > node.keys[i]:
                    i += 1
            
            self._insert_non_full(node.children[i], key)
    
    def _split_child(self, parent, i):
        """分裂子节点"""
        order = self.order
        child = parent.children[i]
        new_child = BPlusTreeNode(is_leaf=child.is_leaf)
        
        # 将child的后半部分移到new_child
        mid = order - 1
        new_child.keys = child.keys[mid + 1:]
        child.keys = child.keys[:mid]
        
        if not child.is_leaf:
            new_child.children = child.children[mid + 1:]
            child.children = child.children[:mid + 1]
        else:
            new_child.next = child.next
            child.next = new_child
        
        # 在parent中插入中间键
        parent.keys.insert(i, child.keys[mid])
        parent.children.insert(i + 1, new_child)
    
    def search(self, key):
        """搜索键值"""
        return self._search_node(self.root, key)
    
    def _search_node(self, node, key):
        """递归搜索节点"""
        i = 0
        while i < len(node.keys) and key > node.keys[i]:
            i += 1
        
        if i < len(node.keys) and key == node.keys[i]:
            return True
        
        if node.is_leaf:
            return False
        
        return self._search_node(node.children[i], key)
    
    def range_query(self, start, end):
        """范围查询"""
        result = []
        leaf = self._find_leaf(self.root, start)
        
        while leaf:
            for key in leaf.keys:
                if start <= key <= end:
                    result.append(key)
                elif key > end:
                    return result
            leaf = leaf.next
        
        return result
    
    def _find_leaf(self, node, key):
        """找到应该包含key的叶节点"""
        if node.is_leaf:
            return node
        
        i = 0
        while i < len(node.keys) and key > node.keys[i]:
            i += 1
        
        return self._find_leaf(node.children[i], key)
    
    def traverse(self):
        """遍历所有键（中序遍历）"""
        result = []
        self._traverse_node(self.root, result)
        return result
    
    def _traverse_node(self, node, result):
        """递归遍历节点"""
        if node.is_leaf:
            result.extend(node.keys)
        else:
            for i in range(len(node.keys)):
                self._traverse_node(node.children[i], result)
                result.append(node.keys[i])
            self._traverse_node(node.children[-1], result)


# 测试示例
if __name__ == "__main__":
    print("=" * 50)
    print("B+树实现 (B+ Tree)")
    print("=" * 50)
    
    bpt = BPlusTree(order=3)
    
    # 插入数据
    values = [10, 20, 5, 6, 12, 30, 7, 17]
    print(f"\n插入数据: {values}")
    for val in values:
        bpt.insert(val)
    
    print(f"中序遍历结果: {bpt.traverse()}")
    
    # 搜索测试
    print("\n搜索测试:")
    for key in [6, 15, 30]:
        result = "找到" if bpt.search(key) else "未找到"
        print(f"  查找 {key}: {result}")
    
    # 范围查询
    print("\n范围查询 [10, 20]:")
    result = bpt.range_query(10, 20)
    print(f"  结果: {result}")
    
    print("\nB+树特点:")
    print("  • 所有数据存储在叶节点")
    print("  • 叶节点形成有序链表")
    print("  • 适合磁盘存储和范围查询")
    print("  • 查找、插入、删除: O(log n)")
    print("  • 应用于数据库索引、文件系统")
